pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java

```diff
@@ -360,4 +368,35 @@ private static CreateMode getCreateMode(EnumSet<CreateOption> options) {
     public long getZkSessionId() {
         return zkc.getSessionId();
     }
+
+    @Override
+    public CompletableFuture<Void> initializeCluster() {
+        if (this.metadataURL == null) {
+            return FutureUtil.failedFuture(new MetadataStoreException("metadataURL is not set"));
+        }
+        if (this.metadataStoreConfig == null) {
+            return FutureUtil.failedFuture(new MetadataStoreException("metadataStoreConfig is not set"));
+        }
+        int chrootIndex = metadataURL.indexOf("/");
+        if (chrootIndex > 0) {
+            String chrootPath = metadataURL.substring(chrootIndex);
+            String zkConnectForChrootCreation = metadataURL.substring(0, chrootIndex);
+            try (ZooKeeper chrootZk = ZooKeeperClient.newBuilder()
+                    .connectString(zkConnectForChrootCreation)
+                    .sessionTimeoutMs(metadataStoreConfig.getSessionTimeoutMillis())
+                    .connectRetryPolicy(
+                            new BoundExponentialBackoffRetryPolicy(metadataStoreConfig.getSessionTimeoutMillis(),
+                                    metadataStoreConfig.getSessionTimeoutMillis(), 0))
+                    .build()) {
+                if (chrootZk.exists(chrootPath, false) == null) {
+                    ZkUtils.createFullPathOptimistic(chrootZk, chrootPath, new byte[0], ZooDefs.Ids.OPEN_ACL_UNSAFE,
+                            CreateMode.PERSISTENT);
+                    log.info("Created zookeeper chroot path {} successfully", chrootPath);
+                }
+            } catch (Exception e) {
+                return FutureUtil.failedFuture(e);
+            }
+        }
+        return CompletableFuture.completedFuture(null);
```

Comment:
IMO there's no need to return a future because it's not an asynchronous API.

Reply:
I mean `MetadataStoreLifecycle#initializeCluster` should not be an asynchronous API. Otherwise it should be renamed to `initializeClusterAsync` and the associated synchronous API `initializeCluster` needs to be added as well.

Reply:
Understood. It seems [`MetadataStore`](https://github.com/apache/pulsar/blob/master/pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataStore.java#L53) APIs are following a different pattern: all IO operations are asynchronous by default. So I think `MetadataStoreLifecycle#initializeCluster` should be designed in the same way.

Reply:
Apart from the creating of the ZooKeeper client you could use all ZK async API calls in order to make it fully async.
But I believe it is overkilling, as you usually are going to wait for this setup operation to complete in real world operations.

I am fine to keep using CompletableFuture, it makes sense to me, for consistency as @fantapsody said

Reply:
I agree with @eolivelli that if you're going to design an asynchronous API, it's better keep using CompletableFuture. I suggested using a synchronous API here just because your implementation is a synchronous way @fantapsody

Reply:
It's true. The synchronous implementation is because I want to keep this piece of code simple and reuse the previous code, and the asynchronous API definition is to leave room for future optimizations if needed.